stew/base10.nim

```diff
@@ -0,0 +1,180 @@
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+## This module implements BASE10 (decimal) encoding and decoding procedures.
+##
+## Encoding procedures are adopted versions of C functions described here:
+## # https://www.facebook.com/notes/facebook-engineering/three-optimization-tips-for-c/10151361643253920
+import results
+export results
+
+{.push raises: [Defect].}
+
+type
+  Base10* = object
+
+{.push overflowChecks: off.}
+proc decode*[A: byte|char](B: typedesc[Base10], T: typedesc[SomeUnsignedInt],
+                           src: openarray[A]): Result[T, cstring] =
+  ## Convert base10 encoded string or array of bytes to unsigned integer.
+  const
+    MaxValue = T(high(T) div 10)
+    MaxNumber = T(high(T) - MaxValue * 10)
+
+  if len(src) == 0:
+    return err("Missing decimal value")
+  var v = T(0)
+  for i in 0 ..< len(src):
+    let d =
+      when A is char:
+        if src[i] >= char(0x30'i8) and src[i] <= char(0x39'i8):
```

Comment:
`let sc = src[i]` to avoid the out-of-bounds checking on every access

Reply:
it will likely be optimized away, but no point in making the compiler work when it can easily be avoided